src/zkevm_specs/evm_circuit/execution/error_write_protection.py

```diff
@@ -0,0 +1,73 @@
+from zkevm_specs.evm_circuit.table import CallContextFieldTag
+
+from ...util import FQ
+from ..instruction import Instruction
+from ..opcode import Opcode
+
+
+# There are some op codes which modify state.
+# There are `[SSTORE, CREATE, CREATE2, CALL, SELFDESTRUCT, LOG0, LOG1, LOG2, LOG3, LOG4]`.
+# When execution call context is read only (static call) and internal call,
+# these op codes running will encounter write protection error
+def error_write_protection(instruction: Instruction):
+    # retrieve op code associated to oog constant error
+    opcode = instruction.opcode_lookup(True)
+    (
+        is_sstore,
+        is_create,
+        is_create2,
+        is_call,
+        is_selfdestruct,
+        is_log0,
+        is_log1,
+        is_log2,
+        is_log3,
+        is_log4,
+    ) = instruction.multiple_select(
+        opcode,
+        (
+            Opcode.SSTORE,
+            Opcode.CREATE,
+            Opcode.CREATE2,
+            Opcode.CALL,
+            Opcode.SELFDESTRUCT,
+            Opcode.LOG0,
+            Opcode.LOG1,
+            Opcode.LOG2,
+            Opcode.LOG3,
+            Opcode.LOG4,
+        ),
+    )
+
+    # Spec 1. opcode must be [SSTORE, CREATE, CREATE2, CALL, SELFDESTRUCT, LOG0, LOG1, LOG2, LOG3, LOG4].
+    instruction.constrain_equal(
+        is_sstore
+        + is_create
+        + is_create2
+        + is_call
+        + is_selfdestruct
+        + is_log0
+        + is_log1
+        + is_log2
+        + is_log3
+        + is_log4,
+        FQ(1),
+    )
+
+    # Spec 2.
+    # current call must be an internal call
+    is_root = instruction.call_context_lookup(CallContextFieldTag.IsRoot)
+    instruction.constrain_equal(is_root, FQ(0))
+    # current call context must be readonly
+    is_static = instruction.call_context_lookup(CallContextFieldTag.IsStatic)
+    instruction.constrain_equal(is_static, FQ(1))
+
+    # Spec 3. for `CALL` op code, `value` must be  non-zero
+    if is_call == FQ(1):
+        instruction.stack_pop()
+        instruction.stack_pop()
+        value = instruction.stack_pop()
```

Comment:
You could use `instruction.stack_lookup` here to directly lookup at the 3rd position in the stack (like a peek), and this way you skip 2 extra lookups.

Reply:
fixed in [e81d764](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/462/commits/e81d764ed77cd2a557903a0936f11a5a465f29bc)